docs/source/packing_slicing_batching.md

```diff
@@ -0,0 +1,56 @@
+<!--- Copyright (c) 2024, NVIDIA CORPORATION.
+SPDX-License-Identifier: BSD-3-Clause -->
+
+# Packing (Slicing and Batching)
+
+If you need to pack samples efficiently, e.g. for language models where padding can be reduced
+easily, this section is for you.
+
+## Concept of efficient Slicing during Batching
+
+The task encoder offer the method {py:meth}`slice_batch <megatron.energon.TaskEncoder.slice_batch>`,
+which gets the samples of a batch and returns a new list of multiple batches. Internally,
+never more than {py:func}`batch_size <megatron.energon.get_train_dataset>` samples are kept, so
+after each yielded batch-slice, new samples will be fetched from the inner dataset to avoid lag.
+
+It is important, to mark custom functions like `encode_sample` as `@stateless` to allow saving
+samples for slicing. If augmentations happen, it should be marked with
+`@stateless(restore_seeds=True)`, to deterministically set the seeds based on the
+`TaskEncoder.current_sample_index`.
+
+Example for padding for a large language model extending the example from the [Task Encoders](task_encoders) section:
+
+```python
+class PackingCaptioningTaskEncoder(CaptioningTaskEncoder):
+    """This class extends the CaptioningTaskEncoder and adds slice_batch for recomputing the batch
+    slice efficiently on-the-fly.
+    Set the batch_size of the get_(train|val)_dataset to an accordingly large number to get a
+    properly sized input samples buffer with some choice.
+    """
+
+    @stateless(restore_seeds=True)
+    def encode_sample(self, ...):
+        # Added `stateless` decorator to allow saving samples for slicing. Will set the seed
+        # deterministically based on the self.current_sample_index.
+        ...
+
+    def slice_batch(self, samples: List[CaptioningSample]) -> List[List[CaptioningSample]]:
```

Comment:
As discussed we might want to change how slicing and batching interact. We should make sure that we can use a batch_size > 1 even when using slicing (on top).